ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala

```diff
@@ -0,0 +1,707 @@
+        template = BranchingSignatories(false, alice, bob)
+        create <- ledger.submitAndWaitRequest(bob, template.create.command)
+        transaction <- ledger.submitAndWaitForTransaction(create)
+        transactions <- ledger.flatTransactions(alice)
+      } yield {
+        assert(transactions.find(_.transactionId != transaction.transactionId).isEmpty)
+      }
+    }
+
+  private[this] val discloseCreateToSignatory =
+    LedgerTest("TXDiscloseCreateToSignatory", "Disclose create to chosen branching controller") {
+      ledger =>
+        for {
+          Vector(alice, bob, eve) <- ledger.allocateParties(3)
+          template = BranchingControllers(alice, true, bob, eve)
+          _ <- ledger.create(alice, template)
+          transactions <- ledger.flatTransactions(alice)
+        } yield {
+          assert(template.arguments == transactions.head.events.head.getCreated.getCreateArguments)
+        }
+    }
+
+  private[this] val notDiscloseCreateToNonChosenBranchingController =
+    LedgerTest(
+      "TXNotDiscloseCreateToNonChosenBranchingController",
+      "Not disclose create to non-chosen branching controller") { ledger =>
+      for {
+        Vector(alice, bob, eve) <- ledger.allocateParties(3)
+        template = BranchingControllers(alice, false, bob, eve)
+        create <- ledger.submitAndWaitRequest(alice, template.create.command)
+        transaction <- ledger.submitAndWaitForTransaction(create)
+        transactions <- ledger.flatTransactions(bob)
+      } yield {
+        assert(transactions.find(_.transactionId != transaction.transactionId).isEmpty)
+      }
+    }
+
+  private[this] val discloseCreateToObservers =
+    LedgerTest("TXDiscloseCreateToObservers", "Disclose create to observers") { ledger =>
+      for {
+        Vector(alice, bob, eve) <- ledger.allocateParties(3)
+        observers = Seq(bob, eve)
+        template = WithObservers(alice, Primitive.List(observers: _*))
+        create <- ledger.submitAndWaitRequest(alice, template.create.command)
+        transactionId <- ledger.submitAndWaitForTransactionId(create)
+        transactions <- ledger.flatTransactions(observers: _*)
+      } yield {
+        assert(transactions.exists(_.transactionId == transactionId))
+      }
+    }
+
+  private[this] val unitAsArgumentToNothing =
+    LedgerTest("TXUnitAsArgumentToNothing", "DAML engine returns Unit as argument to Nothing") {
+      ledger =>
+        for {
+          party <- ledger.allocateParty()
+          template = NothingArgument(party, Primitive.Optional.empty)
+          create <- ledger.submitAndWaitRequest(party, template.create.command)
+          transaction <- ledger.submitAndWaitForTransaction(create)
+        } yield {
+          val contract = assertSingleton("UnitAsArgumentToNothing", createdEvents(transaction))
+          assertEquals("UnitAsArgumentToNothing", contract.getCreateArguments, template.arguments)
+        }
+    }
+
+  private[this] val agreementText =
```

Comment:
I've added an agreement text to the `Dummy` template, so it is easier to create create a contract and observe the agreement text. I needed this for an additional test for the `ActiveContractsService`.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/2713/commits/edbec5015d7c47dfd557c34af94900a47f03e2d0